File: src/checkers-cpp/Move.cpp
#include "Move.h"


Move::Move()
{
}

Move::Move(const Move & move)
{
    this->seq = move.seq;
}

Move::Move(const vector<Position> seq)
{
    this->seq = seq;
}

Move::Move(const string & input)
{
    int len = input.length();
    if (input == "-1"){
        return;
    }

    string delimiter = "-";
    vector<string> points = split(input,delimiter);
    for (int i = 0; i<points.size();++i)
    {
        string point = points[i];
        string result;
        for (int j = 1; j<point.length()-1;++j)
        {
            result += point[j];
        }
        vector<string> xy = split(result,",");
        int x = stoi(xy[0]);
        int y = stoi(xy[1]);
        Position coordinate{x,y};
        seq.push_back(coordinate);


    }

}

vector<string> Move::split(string input,const string delimiter){
    vector<string> result;
    size_t pos = 0;
    string token;
    while ((pos = input.find(delimiter)) != string::npos) {
        token = input.substr(0, pos);
        result.push_back(token);
        input.erase(0, pos + delimiter.length());
    }
    result.push_back(input);
    return result;
}

string Move::toString()
{
    string result;
    for (int i = 0;i<seq.size();++i)
    {
        result += "("+to_string(seq[i][0])+","+to_string(seq[i][1])+")";
        if (i != seq.size()-1)
        {
            result += "-";
        }
    }
    return result;
}

bool Move::isCapture()
{
    if (this->seq.size()>2)
        return true;
    return abs(seq[0][0]-seq[1][0]) > 1;
}

File: src/checkers-java/Move.java
import java.util.Vector;

/*
This module has the Move Class which is the class which handles moves on the board.

We are following the javadoc docstring format which is:
@param Tag Describes the input parameters of the function
@return Description of what the function returns
@throws Tag Describes the errors this function can raise
 */

/**
 * This class is used to describe the moves being made on the board.
 */
public class Move {
	Vector<Position> seq = new Vector<Position>();
	public boolean isCapture = false;

	/**
	 * Initializes Move Object
	 * @param m a sequence of position that the checker pieces will take during the execution of this move
	 *               |  |
	 *             --------
	 *               | X|
	 *             --------
	 *               |  |
	 *             --------
	 *               | X|
	 *             ________
	 *             O |  |
	 *
	 *         In the example above, m should be [(0,0),(2,2),(0,4)]
	 */
	public Move( Move m )
	{
		this.seq = m.seq;
	}

	/**
	 * Initializes Move Object
	 * @param s a sequence of position that the checker pieces will take during the execution of this move
	 */
	public Move( Vector<Position> s)
	{
		this.seq = s;
	}

	/**
	 * Initializes Move Object
	 * @param input a sequence of position (in String format) that the checker pieces will take during the execution of this move
	 *              a string example would be "0, 0 - 2, 2 - 0, 4", given the example above
	 */
	public Move(String input)
	{
		//int len = input.length();
		if(input.equals("-1"))
		{
			seq = new Vector<Position>();
		}else{
	    String delimiter = "-";
	    String[] points = input.split(delimiter);
	    for (int i = 0; i<points.length;++i)
	    {
	        String point = points[i];
	        String result = new String();
	        for (int j = 1; j<point.length()-1;++j)
	        {
	            result = result + point.charAt(j);
	        }
	        String[] xy = result.split(",");
	        int x = Integer.parseInt(xy[0]);
	        int y = Integer.parseInt(xy[1]);
	        Position to_add = new Position(x,y);
	        seq.addElement(to_add);
	    }
		}
	}

	public Vector<String> split(String input,String delimiter) {
	    Vector<String> result = new Vector<String> ();
	    int pos = 0;
	    String token;
	    pos = input.indexOf(delimiter);
	    while (pos != -1)
	    {
	        token = input.substring(0, pos);
	        result.addElement(token);
	        input.replace(input.substring(0, pos+delimiter.length()), "");
		    pos = input.indexOf(delimiter);
	    }
	    result.addElement(input);
	    return result;
	}

	/**
	 * Generates a string for this Move object
	 * @return a string that represents this Move object
	 */
	public String toString()
	{
	    String result = new String();
	    for (int i = 0;i<seq.size();++i)
	    {
	        result += "("+Integer.toString(seq.elementAt(i).getX())+","+Integer.toString(seq.elementAt(i).getY())+")";
	        if (i != seq.size()-1)
	        {
	            result += "-";
	        }
	    }
	    return result;
	}
}
